### ISSUE TITLE
Example server & client teardown

### ISSUE THREAD

**Issue by davidhassel** [aiortc/aioquic#64]
When running the given example client and server, it does not appear that the connection teardown is done properly. This can be seen on the server side when running in verbose mode. Several loss detections are triggered and ping probe packets are sent, when the client have actually successfully finished. Any thoughts?

**Comment by jlaine** [comment]
Can you be a bit more specific about what is wrong? Do you not see a CONNECTION CLOSE frame?

**Comment by davidhassel** [comment]
Server side:

`python3 examples/http3_server.py --certificate tests/ssl_cert.pem --private-key tests/ssl_key.pem -v
2020-02-19 09:56:51,042 DEBUG asyncio Using selector: EpollSelector
2020-02-19 09:56:54,221 DEBUG quic [a91d5a923fae32b0] Network path ('::ffff:127.0.0.1', 56106, 0, 0) discovered
2020-02-19 09:56:54,222 DEBUG quic [a91d5a923fae32b0] QuicConnectionState.FIRSTFLIGHT -> QuicConnectionState.CONNECTED
2020-02-19 09:56:54,228 DEBUG quic [a91d5a923fae32b0] TLS State.SERVER_EXPECT_CLIENT_HELLO -> State.SERVER_EXPECT_FINISHED
2020-02-19 09:56:54,232 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.INITIAL
2020-02-19 09:56:54,233 DEBUG quic [a91d5a923fae32b0] Network path ('::ffff:127.0.0.1', 56106, 0, 0) validated by handshake
2020-02-19 09:56:54,238 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.INITIAL
2020-02-19 09:56:54,238 DEBUG quic [a91d5a923fae32b0] TLS State.SERVER_EXPECT_FINISHED -> State.SERVER_POST_HANDSHAKE
2020-02-19 09:56:54,239 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.HANDSHAKE
2020-02-19 09:56:54,239 INFO quic [a91d5a923fae32b0] ALPN negotiated protocol h3-25
2020-02-19 09:56:54,239 DEBUG quic [a91d5a923fae32b0] Stream 2 created by peer
2020-02-19 09:56:54,239 DEBUG quic [a91d5a923fae32b0] Stream 6 created by peer
2020-02-19 09:56:54,239 DEBUG quic [a91d5a923fae32b0] Stream 10 created by peer
2020-02-19 09:56:54,240 DEBUG quic [a91d5a923fae32b0] Stream 0 created by peer
2020-02-19 09:56:54,291 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:56:54,291 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 10
2020-02-19 09:56:54,367 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:56:54,367 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 11
2020-02-19 09:56:54,517 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:56:54,517 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 12
2020-02-19 09:56:54,817 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:56:54,817 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 13
2020-02-19 09:56:55,415 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:56:55,415 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 14
2020-02-19 09:56:56,612 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:56:56,612 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 15
2020-02-19 09:56:59,004 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:56:59,004 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 16
2020-02-19 09:57:03,787 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:57:03,787 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 17
2020-02-19 09:57:13,351 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:57:13,351 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 18
2020-02-19 09:57:32,481 DEBUG quic [a91d5a923fae32b0] Loss detection triggered
2020-02-19 09:57:32,481 DEBUG quic [a91d5a923fae32b0] Sending PING (probe) in packet 19
2020-02-19 09:57:54,269 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.INITIAL
2020-02-19 09:57:54,269 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.HANDSHAKE
2020-02-19 09:57:54,270 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.ONE_RTT
2020-02-19 09:57:54,270 DEBUG quic [a91d5a923fae32b0] QuicConnectionState.CONNECTED -> QuicConnectionState.TERMINATED
`

Client side:

`python3 examples/http3_client.py --ca-certs tests/pycacert.pem https://localhost:4433/ -v
2020-02-19 09:56:54,214 DEBUG asyncio Using selector: EpollSelector
2020-02-19 09:56:54,220 DEBUG quic [a91d5a923fae32b0] TLS State.CLIENT_HANDSHAKE_START -> State.CLIENT_EXPECT_SERVER_HELLO
2020-02-19 09:56:54,229 DEBUG quic [a91d5a923fae32b0] QuicConnectionState.FIRSTFLIGHT -> QuicConnectionState.CONNECTED
2020-02-19 09:56:54,230 DEBUG quic [a91d5a923fae32b0] TLS State.CLIENT_EXPECT_SERVER_HELLO -> State.CLIENT_EXPECT_ENCRYPTED_EXTENSIONS
2020-02-19 09:56:54,231 DEBUG quic [a91d5a923fae32b0] TLS State.CLIENT_EXPECT_ENCRYPTED_EXTENSIONS -> State.CLIENT_EXPECT_CERTIFICATE_REQUEST_OR_CERTIFICATE
2020-02-19 09:56:54,231 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.INITIAL
2020-02-19 09:56:54,232 DEBUG quic [a91d5a923fae32b0] TLS State.CLIENT_EXPECT_CERTIFICATE_REQUEST_OR_CERTIFICATE -> State.CLIENT_EXPECT_CERTIFICATE_VERIFY
2020-02-19 09:56:54,235 DEBUG quic [a91d5a923fae32b0] TLS State.CLIENT_EXPECT_CERTIFICATE_VERIFY -> State.CLIENT_EXPECT_FINISHED
2020-02-19 09:56:54,236 DEBUG quic [a91d5a923fae32b0] TLS State.CLIENT_EXPECT_FINISHED -> State.CLIENT_POST_HANDSHAKE
2020-02-19 09:56:54,236 INFO quic [a91d5a923fae32b0] ALPN negotiated protocol h3-25
2020-02-19 09:56:54,237 INFO client New session ticket received
2020-02-19 09:56:54,237 DEBUG quic [a91d5a923fae32b0] Stream 3 created by peer
2020-02-19 09:56:54,237 DEBUG quic [a91d5a923fae32b0] Stream 7 created by peer
2020-02-19 09:56:54,237 DEBUG quic [a91d5a923fae32b0] Stream 11 created by peer
2020-02-19 09:56:54,237 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.INITIAL
2020-02-19 09:56:54,240 DEBUG quic [a91d5a923fae32b0] Discarding epoch Epoch.HANDSHAKE
2020-02-19 09:56:54,241 DEBUG quic [a91d5a923fae32b0] Stream 15 created by peer
2020-02-19 09:56:54,250 INFO client Received 1068 bytes in 0.0 s (0.707 Mbps)
`

If you look closer at the timestamp, the client finishes around 09:56:54. At that time the server starts to trigger lost detection, following by terminating the connection 60 seconds later at 09:57:54. The code used is directly from git and not modified.

**Comment by jlaine** [comment]
OK thanks, something does look wrong, I'll definitely have a look when I get a chance.

**Comment by jlaine** [comment]
@davidhassel I am not able to reproduce this issue locally, on every run I've made I get a clean shutdown triggered by the client's `CONNECTION CLOSE`. Could you maybe get a Wireshark capture as I'm not sure why your test differs from mine?

**Comment by jlaine** [comment]
Ping @davidhassel ?

**Comment by davidhassel** [comment]
@jlaine Sorry for the late response. I have attached a wireshark capture, as well as qlog captures. The wireshark capture is used on the latest code from git, while the qlog capture is used on the commit at April 14 (I did not get any qlog file on client side after your update on May 28 to use common logging code).

I used Ubuntu 18.04.4 and have had this problem on several machines.

[aioquic.zip](https://github.com/aiortc/aioquic/files/4757725/aioquic.zip)
